solr/core/src/test/org/apache/solr/core/FSHATest.java

```diff
@@ -0,0 +1,315 @@
+        long individualRequestStart = start;
+        int count = 0;
+        while (nrtCore.equals(hostCore = getHostCoreName(COLL, qaJettyBase, cluster.getSolrClient(), p -> p.add("shards.preference", "replica.type:NRT")))) {
+          count++;
+          individualRequestStart = System.currentTimeMillis();
+        }
+        long now = System.currentTimeMillis();
+        log.info("phase1 NRT queries count={}, overall_duration={}, baseline_expected_overall_duration={}, switch-to-pull_duration={}", count, now - start, establishBaselineMs, now - individualRequestStart);
+        // default tolerance of 500ms below should suffice. Failover to PULL for this case should be very fast,
+        // because our QA-based client already knows both replicas are active, the index is stable, so the moment
+        // the client finds NRT is down it should be able to failover immediately and transparently to PULL.
+        assertWithinTolerance(establishBaselineMs, now - start);
+        assertEquals("when we break out of the NRT query loop, should be b/c routed to PULL", pullCore, hostCore);
+        SolrInputDocument d = new SolrInputDocument();
+        d.addField("id", "345");
+        d.addField("desc_s", "Another Document");
+        // attempts to add another doc while NRT is down should fail, then eventually succeed when NRT comes back up
+        count = 0;
+        start = System.currentTimeMillis();
+        individualRequestStart = start;
+        for ( ; ; ) {
+          try {
+            client.add(COLL, d);
+            client.commit(COLL);
+            break;
+          } catch (SolrException ex) {
+            // we expect these until nrtJetty is back up.
+            count++;
+            Thread.sleep(100);
+          }
+          individualRequestStart = System.currentTimeMillis();
+        }
+        now = System.currentTimeMillis();
+        log.info("successfully added another doc; duration: {}, overall_duration={}, baseline_expected_overall_duration={}, exception_count={}", now - individualRequestStart, now - start, nrtDowntimeMs, count);
+        // NRT replica is back up, registered as available with Zk, and availability info has been pulled down by
+        // our PULL-replica-based `client`, forwarded indexing command to NRT, index/commit completed. All of this
+        // accounts for the 3000ms tolerance allowed for below. This is not a strict value, and if it causes failures
+        // regularly we should feel free to increase the tolerance; but it's meant to provide a stable baseline from
+        // which to detect regressions.
+        assertWithinTolerance(nrtDowntimeMs, now - start, 3000);
+        count = 0;
+        start = System.currentTimeMillis();
+        individualRequestStart = start;
+        while (pullCore.equals(hostCore = getHostCoreName(COLL, qaJettyBase, cluster.getSolrClient(), p -> {
+          p.set(CommonParams.Q, "id:345");
+          p.add("shards.preference", "replica.type:NRT");
```

Comment:
it's optional to specify "pull" replica here ? 

CCing @magibney @noblepaul

Reply:
I'm not sure I understand the question. You could indeed specify `shards.preference=replica.type:PULL`, but this test is specifically exercising our expected use-case, where in the request we'll _always_ be asserting a preference for NRT replicas. We're just testing assumptions about what happens when our preference cannot be satisfied (i.e., the request should fallback to PULL replicas).

Reply:
Thanks. Just wanted to verify this 
```
(i.e., the request should fallback to PULL replicas).
```

We pass both the preference in following way. Tested locally
```
"shards.preference", "replica.type:NRT,replica.type:PULL")
```

Reply:
ohhh yes! I see what you mean. Those should be functionally equivalent. Sorry I misunderstood the question!